docs/docs/schema-customization.md

```diff
@@ -0,0 +1,591 @@
+  for all fields that should be available for querying:
+
+```js:title=gatsby-node.js
+exports.sourceNodes = ({ actions }) => {
+  const { createTypes } = actions
+  const typeDefs = `
+    type AuthorJson implements Node @dontInfer {
+      name: String!
+      firstName: String!
+      email: String!
+      joinedAt: Date
+    }
+  `
+  createTypes(typeDefs)
+}
+```
+
+Note that you don't need to explicitly provide the Node interface fields (`id`,
+`parent`, etc.), Gatsby will automatically add them for you.
+
+> If you wonder about the exclamation marks - those allow specifying nullability
+> in GraphQL, i.e. if a field value is allowed to be `null` or not.
+
+### Nested types
+
+So far we have only been dealing with scalar values (`String` and `Date`;
+GraphQL also knows `ID`, `Int`, `Float`, `Boolean` and `JSON`). Fields can
+however also contain complex object values. To target those fields in SDL, you
+can provide a full type definition for the nested type, which can be arbitrarily
+named (as long as the name is unique in the schema). In our example project, the
+`frontmatter` field on the `MarkdownRemark` node type is a good example. Say we
+want to ensure that `frontmatter.tags` will always be an array of strings.
+
+```js:title=gatsby-node.js
+exports.sourceNodes = ({ actions }) => {
+  const { createTypes } = actions
+  const typeDefs = `
+    type MarkdownRemark implements Node {
+      frontmatter: Frontmatter
+    }
+    type Frontmatter {
+      tags: [String!]!
+    }
+  `
+  createTypes(typeDefs)
+}
+```
+
+Note that with `createTypes` we cannot directly target a `Frontmatter` type
+without also specifying that this is the type of the `frontmatter` field on the
+`MarkdownRemark` type, The following would fail because Gatsby would have no way
+of knowing which field the `Frontmatter` type should be applied to:
+
+```js:title=gatsby-node.js
+exports.sourceNodes = ({ actions }) => {
+  const { createTypes } = actions
+  const typeDefs = `
+    # This will fail!!!
+    type Frontmatter {
+      tags: [String]!
+    }
+  `
+  createTypes(typeDefs)
+}
+```
+
+It is useful to think about your data, and the corresponding GraphQL schema, by
+alwas starting from the Node types created by source and transformer plugins.
+
+> Note that the `Frontmatter` type must not implement the Node interface since
+> it is not a top-level type created by source or transformer plugins: it has no
+> `id` field, and is just there to describe the data shape on a nested field.
+
+### Gatsby Type Builders
+
+In many cases, GraphQL SDL provides a succinct way to provide type definitions
+for your schema. If however you need more flexibility, `createTypes` also
+accepts `graphql-js` types, as well as definitions provided with the help of
+Gatsby Type Builders, which are more flexible than SDL syntax but less verbose
+than `graphql-js`. They are accessible on the `schema` argument passed to Node
```

Comment:
I feel like mentioning `graphql-js` types together with Gatsby Type Builders is a bit confusing here. This could just talk about Gatsby Type Builders, then mention `graphql-js` types after the code example?

Reply:
:+1: Yes thats better!